### ISSUE TITLE
Deadlock with rmdir and rename

### ISSUE THREAD

**Issue by spectral54** [libfuse/libfuse#695]
The partial locking support in `queue_element_wakeup` can cause a deadlock under the following situation:

- something triggers a two-path queue entry (such as for a `rename` operation) to be the first entry and to partially lock
- an `rmdir` for a common grandparent directory of the two paths in the queue arrives

In this situation, the 'rmdir' attempts to acquire a write lock on the directory it wants to remove. This does not succeed, since the reader lock is already held by the partially-locked rename, so the directory is instead marked as "writer waiting" by adding `TREELOCK_WAIT_OFFSET` (`INT_MIN`) to the `treelock`. Every time afterward that the queue attempts to process sees:
- first queue element (`rename`):
    - one of the paths is already locked, so it's ignored
    - the other path tries to reader lock the directory that `rmdir` wants to remove, but fails because there's a writer waiting
    - (no forward progress on this queue element)
- second queue element (`rmdir`):
    - attempt to lock the directory, it fails due to it already being locked
    - (no forward progress on this queue element due to `treelock` already indicating that there's a writer waiting)
 
I've been able to reproduce this reliably using a modified version of the `hello.c` example
[deadlock.patch.txt](https://github.com/libfuse/libfuse/files/9457305/deadlock.patch.txt), and using the command line (in zsh, but should work in bash too):

```
# Build it and run it
$ mkdir -p /tmp/hello
$ gcc -Wall hello.c `pkg-config fuse3 --cflags --libs` -o hello
$ ./hello /tmp/hello

# Reproduce the issue
$ getfattr -nuser.sleep_5s /tmp/hello/rmthis/foo/file.txt &
$ sleep 3
$ getfattr -nuser.sleep_5s /tmp/hello/rmthis/foo/file.txt &
$ mv /tmp/hello/rmthis/{foo,bar}/file.txt &
$ sleep 3
$ rmdir /tmp/hello/rmthis
```

What this is doing is:
1. create a situation where /rmthis, /rmthis/foo, and /rmthis/foo/file.txt all have _two_ reader locks on them due to the getxattrs [note: on my machine only one `getfattr` call is necessary, but if nothing issues spurious getxattr calls for things like the security xattrs, the `sleep 3` and extra `getfattr` should set up things to execute in the proper order, but that's untested]
2. attempt to rename /rmthis/foo/file.txt -> /rmthis/bar/file.txt. This attempts to grab a write lock on /rmthis/foo/file.txt, fails (it's reader locked), marks /rmthis/foo/file.txt as "writer waiting" [this is not relevant to the issue], unlocks everything, adds it to the queue
3. when some getxattr call that has /rmthis/foo/file.txt reader locked completes (either the first one we started in the background, or one of the ones for the 'security' attributes) and there's a queue, it tries to process the queue, which currently has one element in it: the rename.
    - /rmthis has only reader locks, so /rmthis/foo/file.txt fails to lock (reader lock still held by the second getxattr call), but locking /rmthis/bar/file.txt succeeds. This adds another reader lock on /rmthis (so it's now held by 2 or more readers: the second getxattr, the rename, and maybe other getxattrs).
4. sleep 3s to try to ensure that the first `getfattr` has completed before attempting to do the rmdir (in case there's no extraneous reader locks and queue executions) - we need step 3 of this explanation to have happened, as partial lock acquisition only happens when an item is in the queue, not *before* it gets added to the queue.
5. try to rmdir /rmthis. This attempts to grab a write lock on /rmthis, fails (open with reader lock by the second `getfattr` and by /rmthis/bar/file.txt, and possibly others), marks it as "writer waiting [this time that's important], and adds itself to the queue.
    -The queue is now two elements: `{/rmthis/foo/file.txt [unlocked], /rmthis/bar/file.txt [locked]}`, `/rmthis`
6. the second `getfattr` completes, decrements the reader lock count on /rmthis, but this doesn't completely unlock it. Processes the queue: attempt to lock /rmthis/foo/file.txt (fails because /rmthis can't be locked with a reader lock due to a writer waiting), ignores already locked /rmthis/bar/file.txt. Move on to second element: attempt to lock /rmthis, fails because /rmthis/bar/file.txt has /rmthis open with a reader lock.

Proposed fix: remove this `if`: https://github.com/libfuse/libfuse/blob/master/lib/fuse.c#L1136, disallowing partial locking. I have not yet understood what starvation issue partial locking is solving, so this might not be a good solution? All other possible solutions involved detecting that the queue makes no forward progress and unlocking the first element if it's partially locked and then reprocessing the queue. I'm concerned that wouldn't completely solve the issue if (a) enough other work is arriving such that the queue as a whole makes progress every time it's processed, but these specific elements do not, or (b) nothing else manages to start queue processing (because the queue is only processed when a write operation unlocks - if no mutation operation comes in that *succeeds* [due to the deadlock], then I'm worried the queue will never be processed to detect that it doesn't make "forward progress" and thus try the fallback of unlocking the partial lock).

**Comment by Nikratio** [comment]
Thank you for the report and detailed investigation!

My suggestion would be to look at `git blame` for the relevant code. Hopefully, the commit message will specify why partial locking is being used. If not, then I'm afraid this knowledge is most likely lost in the mists of time.

**Comment by spectral54** [comment]
Looks like it was https://github.com/libfuse/libfuse/commit/66956b6f7a254517332fe6e1fa3efab1676e6239, which doesn't have an explanation of the concern (besides the comment).

The only starvation issue I could identify, since we support reader locks, would be if the lock queue was set up via some artificial mechanism to have the first item be a rename or hard link, and then have a never-ending stream of grab-a-write-lock requests (which I think limits it to link, rename, unlink, rmdir) on _both_ paths, perfectly ordered such that every time we process the queue, at least one of them is held. This seems exceedingly unlikely to happen accidentally, and not trivial to trigger intentionally, while this deadlock I'm describing here is both more severe (deadlock instead of starvation - once deadlocked, you can't recover without restarting the filesystem) and seems like it's more likely to occur accidentally.

I'll craft a change to remove this support.

**Comment by aakefbs** [comment]
@spectral54 I find it hard to follow your bug report. Would it be possible to post stack traces? I think it is much easier to see what is going on by looking what different threads are doing. The pstack script in rhel does a good job for that

/usr/bin/pstack:
```
#!/bin/sh

if test $# -ne 1; then
    echo "Usage: `basename $0 .sh` <process-id>" 1>&2
    exit 1
fi

if test ! -r /proc/$1; then
    echo "Process $1 not found." 1>&2
    exit 1
fi

# GDB doesn't allow "thread apply all bt" when the process isn't
# threaded; need to peek at the process to determine if that or the
# simpler "bt" should be used.

backtrace="bt"
if test -d /proc/$1/task ; then
    # Newer kernel; has a task/ directory.
    if test `/bin/ls /proc/$1/task | /usr/bin/wc -l` -gt 1 2>/dev/null ; then
        backtrace="thread apply all bt"
    fi
elif test -f /proc/$1/maps ; then
    # Older kernel; go by it loading libpthread.
    if /bin/grep -e libpthread /proc/$1/maps > /dev/null 2>&1 ; then
        backtrace="thread apply all bt"
    fi
fi

GDB=${GDB:-gdb}

# Run GDB, strip out unwanted noise.
# --readnever is no longer used since .gdb_index is now in use.
$GDB --quiet -nx $GDBARGS /proc/$1/exe $1 <<EOF 2>&1 | 
set width 0
set height 0
set pagination no
$backtrace
EOF
/bin/sed -n \
    -e 's/^\((gdb) \)*//' \
    -e '/^#/p' \
    -e '/^Thread/p'
```

**Comment by spectral54** [comment]
In the threads below, thread 5 is attempting to rename `/tmp/hello/rmthis/foo/file.txt` to `/tmp/hello/rmthis/bar/file.txt`, while thread 4 is attempting to delete `/tmp/hello/rmthis`.

The nodeids are as follows, and I've put the node->treelock values in parenthesis after the name:
1. / (0)
2. /rmthis (0x80000001: 1 reader, at least one writer waiting for that reader to go away)
3. /rmthis/foo (0)
4. /rmthis/foo/file.txt (0)
5. /rmthis/bar (1: 1 reader)
6. /rmthis/bar/file.txt (-1: held with a write lock)

The lockq is:
1. Rename {/rmthis/foo(nodeid=3)/file.txt [not locked], /rmthis/bar(nodeid=5)/file.txt [locked]}
2. Rmdir {/(nodeid=1)/rmthis [not locked]}
3. Some third element that I don't think is relevant, waiting on /rmthis/foo
4. Some fourth element that I don't think is relevant, waiting on /rmthis/foo

Backtraces:

```
Thread 6 (Thread 0x7ffff5957640 (LWP 2080693)):
#0  __GI___libc_read (nbytes=1052672, buf=0x7ffff5055010, fd=3) at ../sysdeps/unix/sysv/linux/read.c:26
#1  __GI___libc_read (fd=3, buf=0x7ffff5055010, nbytes=1052672) at ../sysdeps/unix/sysv/linux/read.c:24
#2  0x00007ffff7f9f8f9 in fuse_session_receive_buf_int (se=0x555555559890, buf=buf@entry=0x7fffec001258, ch=0x0) at ../lib/fuse_lowlevel.c:2870
#3  0x00007ffff7f9a660 in fuse_do_work (data=0x7fffec001240) at ../lib/fuse_loop_mt.c:126
#4  0x00007ffff7d6dd80 in start_thread (arg=0x7ffff5957640) at pthread_create.c:481
#5  0x00007ffff7e87baf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 5 (Thread 0x7ffff625a640 (LWP 2080677)):
#0  0x00007ffff7d7a0fa in __futex_abstimed_wait_common64 (futex_word=futex_word@entry=0x7ffff6259b80, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0, cancel=cancel@entry=true) at ../sysdeps/nptl/futex-internal.c:74
#1  0x00007ffff7d7a15b in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7ffff6259b80, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0) at ../sysdeps/nptl/futex-internal.c:123
#2  0x00007ffff7d73c30 in __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x555555559508, cond=0x7ffff6259b58) at pthread_cond_wait.c:504
#3  __pthread_cond_wait (cond=0x7ffff6259b58, mutex=0x555555559508) at pthread_cond_wait.c:619
#4  0x00007ffff7f8f203 in wait_path (f=f@entry=0x5555555594a0, qe=qe@entry=0x7ffff6259b50) at ../lib/fuse.c:1204
#5  0x00007ffff7f9070b in get_path2 (f=f@entry=0x5555555594a0, nodeid1=nodeid1@entry=3, name1=name1@entry=0x7ffff5958040 "file.txt", nodeid2=nodeid2@entry=5, name2=name2@entry=0x7ffff5958049 "file.txt", path1=path1@entry=0x7ffff6259c50, path2=0x7ffff6259c58, wnode1=0x7ffff6259c60, wnode2=0x7ffff6259c68) at ../lib/fuse.c:1366
#6  0x00007ffff7f92bd0 in fuse_lib_rename (req=0x7fffe0001030, olddir=3, oldname=0x7ffff5958040 "file.txt", newdir=5, newname=0x7ffff5958049 "file.txt", flags=0) at ../lib/fuse.c:3081
#7  0x00007ffff7f9e803 in fuse_session_process_buf_int (se=0x555555559890, buf=buf@entry=0x7fffec000b78, ch=<optimized out>) at ../lib/fuse_lowlevel.c:2686
#8  0x00007ffff7f9a6d3 in fuse_do_work (data=0x7fffec000b60) at ../lib/fuse_loop_mt.c:162
#9  0x00007ffff7d6dd80 in start_thread (arg=0x7ffff625a640) at pthread_create.c:481
#10 0x00007ffff7e87baf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 4 (Thread 0x7ffff6b5d640 (LWP 2080676)):
#0  0x00007ffff7d7a0fa in __futex_abstimed_wait_common64 (futex_word=futex_word@entry=0x7ffff6b5cbc0, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0, cancel=cancel@entry=true) at ../sysdeps/nptl/futex-internal.c:74
#1  0x00007ffff7d7a15b in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7ffff6b5cbc0, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=0) at ../sysdeps/nptl/futex-internal.c:123
#2  0x00007ffff7d73c30 in __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x555555559508, cond=0x7ffff6b5cb98) at pthread_cond_wait.c:504
#3  __pthread_cond_wait (cond=0x7ffff6b5cb98, mutex=0x555555559508) at pthread_cond_wait.c:619
#4  0x00007ffff7f8f203 in wait_path (f=f@entry=0x5555555594a0, qe=qe@entry=0x7ffff6b5cb90) at ../lib/fuse.c:1204
#5  0x00007ffff7f90818 in get_path_common (f=f@entry=0x5555555594a0, nodeid=nodeid@entry=1, name=name@entry=0x7ffff625b038 "rmthis", path=path@entry=0x7ffff6b5cc68, wnode=wnode@entry=0x7ffff6b5cc60) at ../lib/fuse.c:1227
#6  0x00007ffff7f931d7 in get_path_wrlock (wnode=0x7ffff6b5cc60, path=0x7ffff6b5cc68, name=0x7ffff625b038 "rmthis", nodeid=1, f=0x5555555594a0) at ../lib/fuse.c:3034
#7  fuse_lib_rmdir (req=0x7fffec0013c0, parent=1, name=0x7ffff625b038 "rmthis") at ../lib/fuse.c:3034
#8  0x00007ffff7f9e803 in fuse_session_process_buf_int (se=0x555555559890, buf=buf@entry=0x7ffff00010a8, ch=<optimized out>) at ../lib/fuse_lowlevel.c:2686
#9  0x00007ffff7f9a6d3 in fuse_do_work (data=0x7ffff0001090) at ../lib/fuse_loop_mt.c:162
#10 0x00007ffff7d6dd80 in start_thread (arg=0x7ffff6b5d640) at pthread_create.c:481
#11 0x00007ffff7e87baf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 3 (Thread 0x7ffff7460640 (LWP 2080526)):
#0  0x00007ffff7e4ea35 in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x7ffff745fbc0, rem=rem@entry=0x7ffff745fbc0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:43
#1  0x00007ffff7e53af3 in __GI___nanosleep (req=req@entry=0x7ffff745fbc0, rem=rem@entry=0x7ffff745fbc0) at ../sysdeps/unix/sysv/linux/nanosleep.c:25
#2  0x00007ffff7e53a2a in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x000055555555546a in hello_getxattr (path=<optimized out>, attr=<optimized out>, buf=<optimized out>, size=<optimized out>) at ../example/hello.c:148
#4  0x00007ffff7f97713 in common_getxattr (f=f@entry=0x5555555594a0, req=req@entry=0x7fffe8000df0, ino=ino@entry=4, name=name@entry=0x7ffff6b5e040 "user.sleep_5s", value=value@entry=0x0, size=size@entry=0) at ../lib/fuse.c:3865
#5  0x00007ffff7f97877 in fuse_lib_getxattr (req=0x7fffe8000df0, ino=4, name=0x7ffff6b5e040 "user.sleep_5s", size=0) at ../lib/fuse.c:3891
#6  0x00007ffff7f9e803 in fuse_session_process_buf_int (se=0x555555559890, buf=buf@entry=0x7ffff0000b78, ch=<optimized out>) at ../lib/fuse_lowlevel.c:2686
#7  0x00007ffff7f9a6d3 in fuse_do_work (data=0x7ffff0000b60) at ../lib/fuse_loop_mt.c:162
#8  0x00007ffff7d6dd80 in start_thread (arg=0x7ffff7460640) at pthread_create.c:481
#9  0x00007ffff7e87baf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 2 (Thread 0x7ffff7d63640 (LWP 2080525)):
#0  0x00007ffff7e4ea35 in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x7ffff7d62bc0, rem=rem@entry=0x7ffff7d62bc0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:43
#1  0x00007ffff7e53af3 in __GI___nanosleep (req=req@entry=0x7ffff7d62bc0, rem=rem@entry=0x7ffff7d62bc0) at ../sysdeps/unix/sysv/linux/nanosleep.c:25
#2  0x00007ffff7e53a2a in __sleep (seconds=0) at ../sysdeps/posix/sleep.c:55
#3  0x000055555555546a in hello_getxattr (path=<optimized out>, attr=<optimized out>, buf=<optimized out>, size=<optimized out>) at ../example/hello.c:148
#4  0x00007ffff7f97713 in common_getxattr (f=f@entry=0x5555555594a0, req=req@entry=0x7ffff0001330, ino=ino@entry=4, name=name@entry=0x7ffff7461040 "user.sleep_5s", value=value@entry=0x0, size=size@entry=0) at ../lib/fuse.c:3865
#5  0x00007ffff7f97877 in fuse_lib_getxattr (req=0x7ffff0001330, ino=4, name=0x7ffff7461040 "user.sleep_5s", size=0) at ../lib/fuse.c:3891
#6  0x00007ffff7f9e803 in fuse_session_process_buf_int (se=0x555555559890, buf=buf@entry=0x555555579e18, ch=<optimized out>) at ../lib/fuse_lowlevel.c:2686
#7  0x00007ffff7f9a6d3 in fuse_do_work (data=0x555555579e00) at ../lib/fuse_loop_mt.c:162
#8  0x00007ffff7d6dd80 in start_thread (arg=0x7ffff7d63640) at pthread_create.c:481
#9  0x00007ffff7e87baf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 1 (Thread 0x7ffff7d652c0 (LWP 2080516)):
#0  0x00007ffff7d7a0fa in __futex_abstimed_wait_common64 (futex_word=futex_word@entry=0x7fffffffd6d8, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=<optimized out>, cancel=cancel@entry=true) at ../sysdeps/nptl/futex-internal.c:74
#1  0x00007ffff7d7a15b in __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7fffffffd6d8, expected=expected@entry=0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=<optimized out>) at ../sysdeps/nptl/futex-internal.c:123
#2  0x00007ffff7d7644f in do_futex_wait (sem=sem@entry=0x7fffffffd6d8, abstime=0x0, clockid=0) at sem_waitcommon.c:112
#3  0x00007ffff7d764e8 in __new_sem_wait_slow64 (sem=sem@entry=0x7fffffffd6d8, abstime=0x0, clockid=0) at sem_waitcommon.c:184
#4  0x00007ffff7d76561 in __new_sem_wait (sem=sem@entry=0x7fffffffd6d8) at sem_wait.c:43
#5  0x00007ffff7f9a8d8 in fuse_session_loop_mt_32 (se=0x555555559890, config=config@entry=0x7fffffffd768) at ../lib/fuse_loop_mt.c:331
#6  0x00007ffff7f9975d in fuse_loop_mt_32 (f=f@entry=0x5555555594a0, config=config@entry=0x7fffffffd768) at ../lib/fuse.c:4590
#7  0x00007ffff7fa1f67 in fuse_main_real (argc=<optimized out>, argv=<optimized out>, op=<optimized out>, op_size=<optimized out>, user_data=0x0) at ../lib/helper.c:344
#8  0x00005555555551c0 in main (argc=<optimized out>, argv=0x7fffffffd918) at ../example/hello.c:199
```

**Comment by aakefbs** [comment]
I'm a bit confused, why is pthread_mutex_lock(&f->lock) not holding back one thread here?

Thread 5: fuse_lib_rename -> get_path2 -> pthread_mutex_lock(&f->lock)
Thread 4: get_path_wrlock -> get_path_common -> pthread_mutex_lock(&f->lock)

**Comment by spectral54** [comment]
`f->lock` is only held while acquiring a “lock” (not a pthread based lock)
on the individual nodes (by manipulating the treelock property of the nodes
themselves). The lock that’s acquired in get_path_common is dropped as part
of waiting on the condvar in wait_path (pthread_cond_wait requires that the
lock be held before waiting, drops it while waiting, and reacquires it
before returning)

**Comment by aakefbs** [comment]
Ah right, sorry, I had missed that wait_path passes that lock.

**Comment by Nikratio** [comment]
I believe this issue is fixed by the above commit.
